src/topology/category/Profinite.lean

```diff
@@ -0,0 +1,62 @@
+/-
+Copyright (c) 2020 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import topology.category.CompHaus
+
+/-!
+# The category of Profinite Types
+We construct the category of profinite topological spaces,
+often called profinite sets -- perhaps they could be called
+profinite types in Lean.
+
+The type of profinite topological spaces is called `Profinite`. It has a category
+instance and is a fully faithful subcategory of `Top`. The fully faithful functor
+is called `Profinite_to_Top`.
+
+-- TODO
+1) existence of products, limits(?), finite coproducts
+2) `Profinite_to_Top` creates limits?
+-/
+
+open category_theory
+
+/-- The type of profinite topological spaces. -/
+structure Profinite :=
+(to_Top : Top)
+[is_compact : compact_space to_Top]
+[is_t2 : t2_space to_Top]
+[is_td : totally_disconnected_space to_Top]
+
+namespace Profinite
+
+instance : inhabited Profinite := ⟨{to_Top := { α := pempty }}⟩
+
+instance : has_coe_to_sort Profinite := ⟨Type*, λ X, X.to_Top⟩
+instance {X : Profinite} : compact_space X := X.is_compact
+instance {X : Profinite} : t2_space X := X.is_t2
+instance {X : Profinite} : totally_disconnected_space X := X.is_td
+
+instance category : category Profinite := induced_category.category to_Top
+
+end Profinite
+
+/-- The fully faithful embedding of `Profinite` in `Top`. -/
+@[simps {rhs_md := semireducible}, derive [full, faithful]]
+def Profinite_to_Top : Profinite ⥤ Top := induced_functor _
+
+/-- The fully faithful embedding of `Profinite` in `Top`. -/
+@[simps] def Profinite_to_CompHaus : Profinite ⥤ CompHaus :=
+{ obj := λ X, { to_Top := X.to_Top,
```

Comment:
Since we have the `has_coe_to_sort`, can we change this to just be `to_Top := X`? I think then that `is_compact` and `is_hausdorff` should then be automatically inferred

Reply:
```
/-- The fully faithful embedding of `Profinite` in `Top`. -/
@[simps] def Profinite_to_CompHaus : Profinite ⥤ CompHaus :=
{ obj := λ X, { to_Top := ⟨X, _⟩ },
  map := λ _ _ f, f }
```
does work (what you suggested doesn't work, I guess the coercion goes to Sort but then needs to get back to Top), but with the change above `Profinite_to_CompHaus_to_Top` is no longer `rfl` (and as you know I spent about 30 minutes making it `rfl`). Proving instead that the functors are naturally isomorphic and digging down shows that the goal is
```
X: Profinite
⊢ {α := ↥X, str := X.to_Top.str} ≅ X.to_Top
```
and `iso.refl _` doesn't close this. Is this something to do with bundling? I guess I could go ahead and prove this if you think it's worth it but my impression is that what I currently have is the path of least resistance.

Reply:
I think you're right that what you currently have is the best way to do it. That said, you can still automatically infer the instances:
```
{ obj := λ X, { to_Top := X.to_Top },
```
I also think it's a good idea to have a lemma like
```
@[simp]
lemma coe_to_Top {X : Profinite} : (X.to_Top : Type*) = X :=
rfl
```
to make `dsimp` a bit more powerful with these coercions

Reply:
Oh nice! Thanks. I'll add it to CompHaus as well.